doc/md/7-motoko-style-guide.md

```diff
@@ -0,0 +1,1323 @@
+
+### Classes
+
+- Use `class` to create multiple objects of the same shape.
+
+- Name classes after their conceptual functionality, not their implementation, except when having to distinguish multiple different implementations of the same concept. For example, `OrderedMap` vs `HashMap`.
+
+- Classes are both type definitions and factory functions for objects.
+
+    Do not use classes unless both these roles are intended; use plain type aliases or functions returning an object in other cases.
+
+- Do not overuse classes.
+
+    Use a module defining a plain type and functions on it where appropriate. Use classes only as a means for encapsulating state or behavior.
+
+    Rationale: Objects with methods have disadvantages over simple record types with separate functions (see above).
+
+- If values of a class are meant to be shared, the class needs to provide a pair of `share`/`unshare` methods that convert to/from a sharable representation, for example, as a record.
+
+    :::note
+
+    For immutable classes it may seem more natural to make `unshare` a kind of static function. However, even for immutable ones it may depend on constructor arguments (such as an ordering function), so that the a pattern like `Map(compareInt).unshare(x)` seems appropriate.
+
+    :::
+
+- Avoid overloading classes with too many methods, since that is expensive.
+
+    Restrict to a sufficiently small set of canonical methods. Make less essential methods that can be implemented on top of the canonical methods into functions in the enclosing module.
+
+- Use modules for "static" classes or methods.
+
+### Modules
+
+- Use `module` to group definitions, including types, and create a name space for them.
+
+- Where applicable, name modules after the main type or class they implement or provide functions for.
+
+- Limit each module to a single main concept/type/class or closely entangled family of concepts/types/classes.
+
+### Error handling
+
+- Use `Result` for error handling where applicable.
+  - Use `Result<T, E>` rather than exceptions for recoverable errors.
+  - Use meaningful error types instead of generic `Text`.
+
+    ```motoko no-repl
+    type MyError = { #NotFound; #InvalidInput; #PermissionDenied };
+
+    func safeDivide(a : Int, b : Int) : Result.Result<Int, MyError> {
+      if (b == 0) { return #err (#InvalidInput) };
+      #ok (a / b);
+    };
+    ```
+
+- Use `assert` only for programming errors.
+  - `assert` should be reserved for conditions that should never fail if the program is correct.
+
+    ```motoko no-repl
+    assert(x > 0);  // Only use if x should never be <= 0
+    ```
+
+- Use `debug assert` for development-time checks.
+  - `debug assert` is like `assert`, but only runs in non-release builds.
+  - Useful for checking internal assumptions or invariants without impacting production performance.
+  - All `debug` expressions are ignored when compiled with `moc --release`.
```

Comment:
Good addition! I haven't checked but this info about `assert` should also be included in a separate section about `assert` in general, not just in a style guide. I'd maybe link to that section here
Also there I'd include the `dfx`'s `Release` option because people usually are not using `moc` directly